### PR TITLE
Add typescript definition file.

### PR DESCRIPTION
**Pull request by ducksoupdev**
This is a simple Typescript definition for the API. If anyone is writing typescript for node and using your module then this is useful.

### CODE REVIEW

File: jsonfile.d.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+declare module "jsonfile" {
+	var spaces: number;
+	function readFile(filename: string, options:any, callback:Function);
+	function readFileSync(filename: string, options:any);
+	function writeFile(filename: string, obj:any, options:any, callback:Function);
+	function writeFileSync(filename: string, obj:any, options:any);
+}
```
Comment (LinusU) [inline jsonfile.d.ts]: No newline at end of file

File: test/read-file-sync.test.js [hunk 2]
```diff
@@ -43,7 +43,7 @@ describe('+ readFileSync()', function () {
         jf.readFileSync(file)
       }, function (err) {
         assert(err instanceof Error)
-        assert(err.message.match(file))
+        assert(err.message.match(/somefile\.json/))
```
Comment (LinusU) [inline test/read-file-sync.test.js]: Can you explain why this change is needed? :)

Comment (ducksoupdev) [inline test/read-file-sync.test.js]: This test fails when run on Windows with node v0.12.x.

image

File: jsonfile.d.ts (hunk 1 above)
Comment (ducksoupdev) [inline jsonfile.d.ts]: No problem, this can be added.

File: test/read-file-sync.test.js (hunk 2 above)
Comment (LinusU) [inline test/read-file-sync.test.js]: Have you tried echoing out `err.message` and `file` to see why it doesn't match?

Comment (ducksoupdev) [inline test/read-file-sync.test.js]: Yes, I have also done that.

image

Comment (LinusU) [inline test/read-file-sync.test.js]: Ah, I see, the problem is that the string is coerced into a regular expression.

This should work:

```javascript
assert(err.message.indexOf(file) !== -1)
```

Or even prettier, but I don't think it works on <= 0.12

```javascript
assert(err.message.includes(file))
```

**Comment by jprichardson** [comment]
As much as I'd love to support this, it adds an additional maintenance burden. Why do you think this needs to be in this repo and not in https://github.com/borisyankov/DefinitelyTyped/ where there are active TypeScript users creating these definitions?

File: test/read-file-sync.test.js (hunk 2 above)
Comment (ducksoupdev) [inline test/read-file-sync.test.js]: I'm happy to modify to either, although the former would be better

**Comment by LinusU** [comment]
@jprichardson Can't comment on the typed thing since I don't use TypeScript at the moment (interested in starting thought, it seems cool), but how about fixing the tests on Windows thought. The comment I last made should be a fix for it. The problem is the slashes that windows uses for paths.

**Comment by ducksoupdev** [comment]
Editor support for definitions in node modules is now a popular method of including these rather than having to install separate typings. More details can be found here https://github.com/Microsoft/TypeScript/issues/247. I'm happy to help and maintain the typings for this module as it's a great utility I use frequently?

**Comment by blakeembrey** [comment]
Inline type definitions is the correct way to go, but this file is wrong for that. It needs to be an external module, not an ambient module if you want to use it inside the module. You should also be adding a `typings` field to `package.json`.

**Comment by wind-rider** [comment]
I'd also like to have this functionality. The nice thing about adding it to a npm package instead of uploading it to DefinitelyTyped et al is that the user gets his typings immediately so that he doesn't need to download and update them separately.
There is more documentation now at https://github.com/Microsoft/TypeScript/wiki/Typings-for-npm-packages how it can be included in package.json files.
Basically it comes down to adding a ```"typings": "index.d.ts",``` line in package.json and adding the typings to ```index.d.ts```.

I'm new to TypeScript and I don't know the difference between external modules and ambient modules well (@blakeembrey) - I created a similar file to what @ducksoupdev did.

Please see my version below - but I'm new to TypeScript so maybe another approach is better.

```
declare module "jsonfile" {
    
    export var spaces: number;
  
    /*
     * Asynchronous readFile - Asynchronously reads the entire contents of a file.
     *
     * @param fileName
     * @param options An object with optional {encoding} and {flag} properties.  If {encoding} is specified, readFile returns a string; otherwise it returns a Buffer.
     * @param callback - The callback is passed two arguments (err, data), where data is the contents of the file.
     */
    export function readFile(filename: string, options: { encoding: string; flag?: string; }, callback: (err: NodeJS.ErrnoException, data: any) => void): void;
    /*
     * Asynchronous readFile - Asynchronously reads the entire contents of a file.
     *
     * @param fileName
     * @param options An object with optional {encoding} and {flag} properties.  If {encoding} is specified, readFile returns a string; otherwise it returns a Buffer.
     * @param callback - The callback is passed two arguments (err, data), where data is the contents of the file.
     */
    export function readFile(filename: string, options: { flag?: string; }, callback: (err: NodeJS.ErrnoException, data: any) => void): void;
    /*
     * Synchronous readFile - Synchronously reads the entire contents of a file.
     *
     * @param fileName
     * @param options An object with optional {encoding} and {flag} properties.  If {encoding} is specified, readFileSync returns a string; otherwise it returns a Buffer.
     */
    export function readFileSync(filename: string, options: { encoding: string; flag?: string; }): any;
    /*
     * Synchronous readFile - Synchronously reads the entire contents of a file.
     *
     * @param fileName
     * @param options An object with optional {encoding} and {flag} properties.  If {encoding} is specified, readFileSync returns a string; otherwise it returns a Buffer.
     */
    export function readFileSync(filename: string, options?: { flag?: string; }): any;
    
    export function writeFile(filename: string, obj: any, options: { encoding?: string; mode?: number; flag?: string; }, callback?: (err: NodeJS.ErrnoException) => void): void;
    export function writeFile(filename: string, obj: any, options: { encoding?: string; mode?: string; flag?: string; }, callback?: (err: NodeJS.ErrnoException) => void): void;
    export function writeFileSync(filename: string, obj: any, options?: { encoding?: string; mode?: number; flag?: string; }): void;
    export function writeFileSync(filename: string, obj: any, options?: { encoding?: string; mode?: string; flag?: string; }): void;
  
}

```

**Comment by jprichardson** [comment]
@wind-rider thanks. I'm not opposed to it, I just don't want to maintain it. Is there a tool that can generate these files? Is there a tool to convert Flow definitions to TypeScript?

**Comment by blakeembrey** [comment]
@jprichardson I just created an issue around community maintenance of typings - in summary, @typings/collab could be a way forward for you to have type definitions and keep them up to date. Let me know what you think with https://github.com/typings/typings/issues/322.

**Comment by wind-rider** [comment]
@jprichardson I understand, but I think that as long as you don't use typescript directly (which should be a relatively small change in the case of jsonfile since its codebase is small) there is no auto-generation of typings. However, if you would decide to switch then javascript will be generated from your typescript and you will have the advantages of new javascript functions and strong typing in javascript.

@blakeembrey your proposal also sounds like a good idea! Probably they are experts in creating those typings :)

**Comment by RyanZim** [comment]
FWIW, I'm not in favor of this unless someone on the core team is maintaining the typings. Just my 2 cents.

**Comment by jprichardson** [comment]
Nope: https://github.com/jprichardson/node-jsonfile/issues/71#issuecomment-296515551
